Polish the demobench

mikehearn opened the pull request:
<img width="1315" alt="demobench1" src="https://cloud.githubusercontent.com/assets/971089/25151595/67ff32aa-2486-11e7-8c4a-a5f9781f7d77.png">

<img width="1315" alt="demobench2" src="https://cloud.githubusercontent.com/assets/971089/25151609/7360408a-2486-11e7-8452-b04551d769c4.png">

chrisr3 commented on tools/demobench/src/main/resources/net/corda/demobench/views/NodeTerminalView.fxml:
```diff
@@ -3,24 +3,23 @@
-            <Label fx:id="nodeName" style="-fx-font-size: 40; -fx-text-fill: red;"/>
-            <PropertyLabel fx:id="p2pPort" name="P2P port: "/>
+<VBox visible="false" prefHeight="953.0" prefWidth="1363.0" xmlns="http://javafx.com/javafx/8.0.102" xmlns:fx="http://javafx.com/fxml/1" styleClass="terminal-vbox">
+    <HBox prefHeight="95.0" prefWidth="800.0" styleClass="header">
```
The "Launch Web Server" button is currently squashed so that I cannot read which port number the web server has been launched on. With this new UI, I need to do something like:
```xml
<HBox minHeight="140.0" prefWidth="800.0" styleClass="header">
```
and then assign `prefHeight="138.0"` to all of its children to prevent JavaFX from deforming this header panel.

chrisr3 commented on tools/demobench/src/main/resources/net/corda/demobench/views/NodeTerminalView.fxml:
```diff
@@ -3,24 +3,23 @@
             <PropertyLabel fx:id="balance" name="Balance: "/>
         </VBox>
-        <Pane prefHeight="200.0" prefWidth="200.0" HBox.hgrow="ALWAYS"/>
+        <!--<Pane prefHeight="200.0" prefWidth="200.0"/>-->
```
Remove this line.

chrisr3 commented on tools/demobench/build.gradle:
```diff
@@ -33,10 +33,15 @@ repositories {
     }
 }
 
+compileKotlin {
```
Just for DemoBench? I would have thought that this would be "global".

mikehearn commented on tools/demobench/src/main/kotlin/net/corda/demobench/views/NodeTabView.kt:
```diff
@@ -21,12 +31,10 @@ class NodeTabView : Fragment() {
+        const val textWidth = 465.0
         const val maxNameLength = 15
 
         val integerFormat = DecimalFormat()
```
Yeah, I'll remove the ability to change the port numbers. The case where a node doesn't terminate is a bug in demobench, right? The main reason to kill it off is that in such a situation, the intended users of the tool are unlikely to be able to fix it by quickly switching port numbers around. Better to keep the UI simple and focus on robustness of the implementation.

mikehearn commented:
I've addressed your comments and added a couple more tweaks:

* Refresh bug #450 is fixed
* Launch web server button opens a browser to the root page now (or would do ... my setup seems to be kinda busted)

chrisr3 commented on tools/demobench/src/main/kotlin/net/corda/demobench/views/NodeTabView.kt:
```diff
@@ -264,6 +215,7 @@ class NodeTabView : Fragment() {
         if (config != null) {
             nodeConfigView.isVisible = false
+            nodeTab.graphic = ImageView(flags.get()[model.nearestCity.value.countryCode]).apply { fitWidth = 24.0; isPreserveRatio = true }
             config.install(cordapps)
```
You no longer need this line now that you're setting `nodeTab.graphic` in `launchNode(NodeConfig)`.

chrisr3 commented on tools/demobench/src/main/kotlin/net/corda/demobench/views/NodeTerminalView.kt:
```diff
@@ -46,18 +48,23 @@ class NodeTerminalView : Fragment() {
         nodeName.text = config.legalName
-        p2pPort.value = config.p2pPort.toString()
-        launchWebButton.text = "Launch\nWeb Server\n(Port ${config.webPort})"
+        launchWebButton.text = "Launch\nWeb Server"
```
You don't need to alter the `launchWebButton.text` any more, because this is the default value:
```xml
<Button fx:id="launchWebButton" text="Launch&#10;Web Server" .../>
```

mikehearn commented:
This PR now depends on #557

mikehearn commented:
This PR is also pending on the release of TornadoFX 1.7.3 which should occur tonight. I tested with a snapshot. 1.7.3 fixes the width of the auto-completing combo dropdown.

chrisr3 commented:
![demobench-form](https://cloud.githubusercontent.com/assets/25174078/25188593/7d9a01c4-251e-11e7-965b-e46f78f07fc6.jpeg)
The "Launch Web Server" button still isn't quite wide enough, and it's not opening anything in my browser either. Although I wouldn't expect it to until the web server has booted anyway.

chrisr3 commented:
Interestingly, we seem now to insist on JDK >= 8u112, i.e. `build.gradle`:131
```gradle
if (!JavaVersion.current().java8Compatible)
    throw new GradleException("Corda requires Java 8, please upgrade to at least 1.8.0_112")
```
So the .fxml files should probably be consistent with this too.

gendal commented:
* Not reviewed the code.  
* Love the new look-and-feel/reskinning. Makes a big difference.
* Is this branch buildable and runnable obviously/easily? If so, happy to run it and provide feedback from a user's perspective.
* Oh... Checked and confirmed to my satisfaction that GB.png was a correct depiction of the Union Flag... phew!

chrisr3 commented:
We should be able to remove this paragraph from `demobench.rst` now:
> After switching tabs, it may currently be necessary to click on the new tab panel before its contents update correctly. This is a BUG and we're working on it.

chrisr3 commented:
I notice that the tab is no longer being deleted if the node crashes (as opposed to it exiting normally). This is presumably so that the user can still read anything useful that may have been written to the terminal, but it also leaves the RPC client and the Explorer/DB/WebServer buttons active.

Perhaps:
```kotlin
if (exitCode == 0) {
    nodeTab.requestClose()
} else {
    nodeTerminalView.shutdown()
}
```
where `NodeTerminalView.shutdown()` does something like:
```xml
<HBox fx:id="header" prefHeight="95.0" prefWidth="800.0" styleClass="header">
```
```kotlin
class NodeTerminalView : Fragment() {
    ...
    private val header by fxid<HBox>()
    ...

    fun shutdown() {
        header.isDisable = true
        webServer.close()
        explorer.close()
        viewer.close()
        rpc?.close()
    }
```

mikehearn commented:
@chrisr3 All things fixed, I think. I've made the "launch web server" button open the browser once the server has started up. PTAL.

mikehearn commented:
@gendal I slacked you a link to a download. There are one or two more minor things to adjust but I'll do that in a separate PR.

chrisr3 commented on tools/demobench/src/main/kotlin/net/corda/demobench/views/NodeTerminalView.kt:
```diff
@@ -133,6 +144,29 @@ class NodeTerminalView : Fragment() {
+            log.info("Waiting for web server to start ...")
+            val url = URL("http://localhost:${config.webPort}/")
+            val rateLimiter = RateLimiter.create(1.0)
+            while (true) {
```
What happens if the WebServer _never_ starts, e.g. because of a bad CorDapp? Shouldn't this loop give up eventually?

chrisr3 replied:
What happens if the WebServer _never_ starts, e.g. because of a bad CorDapp? Shouldn't this loop give up eventually?

mikehearn commented:
Great comments, thanks. I've addressed them, PTAL.

chrisr3 commented on tools/demobench/src/main/kotlin/net/corda/demobench/web/WebServer.kt:
```diff
@@ -46,12 +57,19 @@ class WebServer internal constructor(private val webServerController: WebServerC
+                future.catch {
+                    log.info("Waiting for web server for ${config.legalName} to start ...")
+                    waitForStart(config.webPort)
+                    URI("http://localhost:${config.webPort}/")
```
Did you mean to leave this line here? It looks like `waitForStart()` is supposed to provide the return value.

mikehearn replied:
You are correct, good sir! Fixed.

chrisr3 commented:
I'm sure the unused import in `WebServer.kt` will disappear on its own ... ;-)
